usr/share/rear/prep/GNU/Linux/210_include_dhclient.sh

```diff
@@ -4,114 +4,123 @@
 # copy the binaries and config files that we require to use dhclient/dhcpcd
 # on the rescue image (IPv4/IPv6)
 
-define_dhclients_variable()
-{   local x
-    dhclients=()
-    for x in "${DHCLIENT_BIN##*/}" \
-        "${DHCLIENT6_BIN##*/}" \
-        dhcpcd dhclient \
-        dhcp6c dhclient6 ;
-    do
-        [ "x$x" == x ] && continue
-        for d in ${dhclients[@]} ; do
-            [ "x$d" = "x$x" ] && continue 2
-        done
-        dhclients=(${dhclients[@]} "$x")
-    done
-    dhclients=${dhclients[@]}
-}
+# Respect an explicit rejection
+if is_false "$USE_DHCLIENT"; then
+    Log "Disabling DHCP client on the rescue system (USE_DHCLIENT='$USE_DHCLIENT')"
+    return 0
+fi
 
-dhcp_interfaces_active() {
-    local my_dhclients
-    my_dhclients=${dhclients}
-    my_dhclients=${dhclients// /|}
-    ps -e | grep -qEs "[ /]($my_dhclients)"
+# Array of known DHCP clients, which must also be known to skel/default/etc/scripts/system-setup.d/58-start-dhclient.sh
+local dhcp_clients=(dhcpcd dhclient dhcp6c dhclient6)
+
+# DHCLIENT_BIN, if set, specifies the DHCP client binary to use on the rescue system.
+# Must be one of the binaries listed in $dhcp_clients. Default (empty): try to auto-detect.
+# FIXME: DHCLIENT_BIN could be documented in default.conf as needed.
+if [[ -n "$DHCLIENT_BIN" ]]; then
+    has_binary "$DHCLIENT_BIN" || Error "DHCLIENT_BIN='$DHCLIENT_BIN' but such a binary could not be found"
+    [[ "${dhcp_clients[*]}" == *" $DHCLIENT_BIN "* ]] || Error "DHCLIENT_BIN='$DHCLIENT_BIN' is not among known DHCP clients (${dhcp_clients[*]})"
```

Comment:
This test does not work:
```
# dhcp_clients=(dhcpcd dhclient dhcp6c dhclient6)

# for DHCLIENT_BIN in ${dhcp_clients[*]} ; do [[ "${dhcp_clients[*]}" == *" $DHCLIENT_BIN "* ]] || echo "DHCLIENT_BIN='$DHCLIENT_BIN' is not among known DHCP clients (${dhcp_clients[*]})" ; done

DHCLIENT_BIN='dhcpcd' is not among known DHCP clients (dhcpcd dhclient dhcp6c dhclient6)
DHCLIENT_BIN='dhclient6' is not among known DHCP clients (dhcpcd dhclient dhcp6c dhclient6)
```
because it tests with leading and trailing blanks which is magic characters coding.
I suggest the more straifgtforward test:
```
IsInArray "$DHCLIENT_BIN" "${dhcp_clients[@]}" || Error "..."
```

Reply:
You are right, I forgot to add blanks: `[[ " ${dhcp_clients[*]} " == *" $DHCLIENT_BIN "* ]]`

Anyway, I've changed it to use `IsInArray`. Occasionally, I miss a rear library function because I'm just a casual user and there would be so many `lib` sources to look at. In this case, I just didn't remember. However, I may deliberately choose not to use a lib function if it does not add sufficient value to compensate a major downside: These functions make manual unit testing harder. I tend to write test beds like this and it can be hard to pull in ReaR lib files due to dependency hell:

```
#!/bin/bash

. usr/share/rear/lib/global-functions.sh

function has_binary () {
    for bin in $@ ; do
        # Suppress success output via stdout which is crucial when has_binary is called
        # in other functions that provide their intended function results via stdout
        # to not pollute intended function results with intermixed has_binary stdout
        # (e.g. the RequiredSharedObjects function) but keep failure output via stderr:
        type $bin 1>/dev/null && return 0
    done
    return 1
}

function IsInArray() {
    element="$1"
    shift
    [[ " $* " == *" $element "* ]]
}

function Log() {
    echo "Log: $*"
}

function Error() {
    echo "Error: $*"
    exit 1
}

function test() {
    . usr/share/rear/prep/GNU/Linux/210_include_dhclient.sh
}

#set -x

ROOTFS_DIR=.

test

echo "$ROOTFS_DIR/etc/rear/rescue.conf:"
cat "$ROOTFS_DIR/etc/rear/rescue.conf"
rm "$ROOTFS_DIR/etc/rear/rescue.conf"
```